torch/csrc/distributed/c10d/init.cpp

```diff
@@ -1527,11 +1549,49 @@ static const auto ProcessGroupNCCLOptionsTorchBind =
 
 static const auto ProcessGroupNCCLTorchBind =
     torch::class_<::c10d::ProcessGroupNCCL>("dist_c10d", "ProcessGroupNCCL")
-        .def(torch::init<
-             const c10::intrusive_ptr<::c10d::Store>&,
-             int64_t,
-             int64_t,
-             const c10::intrusive_ptr<::c10d::ProcessGroupNCCL::Options>&>())
+        .def_pickle(
+            [](const c10::intrusive_ptr<::c10d::ProcessGroupNCCL>& self) {
+              auto base_process_group =
+                  static_intrusive_pointer_cast<::c10d::ProcessGroup>(self);
+              auto name =
+                  ::c10d::DistributedC10d::get()->getNameOfProcessGroup(self);
+              return std::vector<std::string>{name};
+            },
+            [](std::vector<std::string> state) {
+              TORCH_CHECK(
+                  state.size() == 1,
+                  "Expecting exactly 1 state when restoring ProcessGroupNCCL, got: ",
+                  state.size());
+              const auto& process_group_name = state.front();
+              auto base_process_group =
+                  ::c10d::DistributedC10d::get()->getProcessGroupByName(
+                      process_group_name);
+              TORCH_CHECK(
```

Comment:
seems like we already check this inside `getProcessGroupByName`, so this check is likely a code that can't be reached, maybe merge this two check together?